### PR TITLE
Add support for Linux's abstract sockets

### LINKED ISSUE #2526: Support for Linux's abstract sockets
Linux supports listening on an abstract socket that is not represented by a path on the filesystem (https://man7.org/linux/man-pages/man7/unix.7.html). I think ruby's `UNIXSocket` support abstract sockets, but some of the `File.exist?` type checks in puma fail because abstract sockets start with a NUL byte `"\0"`. Thoughts on supporting abstract sockets?

### PR DESCRIPTION
**Pull request by MSP-Greg**
### Description

Closes #2526

Tests for abstract UNIXSockets exist in the updated test framework. Will open a PR for it after all the library file changes have been accepted.  Tested both server & control binding...

### Your checklist for this pull request
- [x] I have reviewed the [guidelines for contributing](../blob/master/CONTRIBUTING.md) to this repository.
- [x] I have added an entry to [History.md](../blob/master/History.md) if this PR fixes a bug or adds a feature. If it doesn't need an entry to HISTORY.md, I have added `[changelog skip]` or `[ci skip]` to the pull request title.
- [ ] I have added (or updated) appropriate tests if this PR fixes a bug or adds a feature.
- [x] My pull request is 100 lines added/removed or less so that it can be easily reviewed.
- [ ] If this PR doesn't need tests (docs change), I added `[ci skip]` to the title of the PR.
- [x] If this closes any issues, I have added "Closes `#issue`" to the PR description or my commit messages.
- [ ] I have updated the documentation accordingly.
- [x] All new and existing tests passed, including Rubocop.

### CODE REVIEW

**Comment by MSP-Greg** [comment]
Rebased and added two simple tests.  The new test framework has a test or two creating hundreds of abstract unix clients...

**Comment by MSP-Greg** [comment]
Thought I'd rebase this.  Also, curl can be used to verify the PR.  Compile the Puma repo, then

```bash
# assumes tcp 127.0.0.1:40000 is avaiable, change as needed
bundle exec ruby -Ilib bin/puma -q -w2 -t5:5 -b tcp://127.0.0.1:40000 test/rackup/realistic_response.ru
curl -kso /dev/null -w '%{size_download}' http://127.0.0.1:40000

bundle exec ruby -Ilib bin/puma -q -w2 -t5:5 -b unix://@puma.unix test/rackup/realistic_response.ru
curl -kso /dev/null -w '%{size_download}' --abstract-unix-socket puma.unix http:

```
The file realistic_response.ru actually returns a 200,000 byte body.  Both curl commands should output 200000, which is the body byte size...

File: lib/puma.rb [hunk 1]
```diff
@@ -39,6 +39,20 @@ def self.ssl?
     HAS_SSL
   end
 
+  def self.abstract_unix_socket?
```
Comment (nateberkopec) [inline lib/puma.rb:42]: Should this be in `detect`?

File: lib/puma/binder.rb [hunk 2]
```diff
@@ -407,24 +412,24 @@ def inherit_unix_listener(path, fd)
...
-      redirects = listeners.map { |a| [a[1].to_i, a[1].to_i] }.to_h
+      redirects = @listeners.map { |a| [a[1].to_i, a[1].to_i] }.to_h
       redirects[:close_others] = true
       redirects
     end
 
     # @version 5.0.0
     def redirects_for_restart_env
-      listeners.each_with_object({}).with_index do |(listen, memo), i|
+      @listeners.each_with_object({}).with_index do |(listen, memo), i|
```
Comment (nateberkopec) [inline lib/puma/binder.rb:432]: Why do we have to work around the accessor method?

File: lib/puma/control_cli.rb [hunk 3]
```diff
@@ -176,7 +176,8 @@ def send_request
         when 'tcp'
           TCPSocket.new uri.host, uri.port
         when 'unix'
-          UNIXSocket.new "#{uri.host}#{uri.path}"
+          UNIXSocket.new(@control_url.start_with?('unix://@') ?
```
Comment (nateberkopec) [inline lib/puma/control_cli.rb:179]: Everywhere that we check for a socket starting with `@`, I think we should have a comment explaining this a check for an abstract socket.

File: lib/puma/binder.rb [hunk 4]
```diff
@@ -374,7 +381,7 @@ def add_unix_listener(path, umask=nil, mode=nil, backlog=1024)
           end
         end
 
-        s = UNIXServer.new(path)
+        s = UNIXServer.new path.sub(/\A@/, "\0")
```
Comment (nateberkopec) [inline lib/puma/binder.rb:384]: Everywhere that we check for a socket starting with `@`, I think we should have a comment explaining this a check for an abstract socket.

File: lib/puma/binder.rb [hunk 5]
```diff
@@ -177,11 +177,19 @@ def parse(binds, logger, log_msg = 'Listening')
           @listeners << [str, io] if io
         when "unix"
           path = "#{uri.host}#{uri.path}".gsub("%20", " ")
+          abstract = false
+          if str.start_with? 'unix://@'
```
Comment (nateberkopec) [inline lib/puma/binder.rb:181]: (this one is explained by the next line so no comment required here)

**Comment by nateberkopec** [comment]
Minor style concerns, otherwise LGTM

File: lib/puma/binder.rb (hunk 2 above)
Comment (MSP-Greg) [reply lib/puma/binder.rb:432]: This is something that is somewhat personal preference, as I doubt there's any performance difference in the general case.

First, `@listeners` is initialized.  Secondly, I haven't looked at it for a while (Puma 6 is not around the corner), but there are a lot of public methods/accessors that could be removed.  Changing our code to allow that is helpful.

### General

1. Removing accessors may break tests, but conversely, accessors added just to make testing simple may not be a good idea.  Tests can use `#instance_variable_get` if needed.
2. Reducing the public API often shows where encapsulation has been broken, often where an accessor should be replaced  with one or more 'action' methods.
3. I'll never forget breaking Capybara.
4. I'm not sure if I changed all the code in Binder to use `@listeners`.

File: lib/puma.rb (hunk 1 above)
Comment (MSP-Greg) [reply lib/puma.rb:42]: I could make it more clear, but I thought `detect` would be for OS dependent constants/methods, and Puma was used for constants/methods that could only be determined after something was required.

To check 'ssl', 'puma/puma_http11' needs to be required.  To check `UNIXSocket` (and abstract), we need to require 'socket'?

File: lib/puma/binder.rb (hunk 4 above)
Comment (MSP-Greg) [reply lib/puma/binder.rb]: Done

File: lib/puma/control_cli.rb (hunk 3 above)
Comment (MSP-Greg) [reply lib/puma/control_cli.rb:180]: Done

**Pull request merged**